File: rin.Framework/Graphics/Descriptors/DescriptorLayoutFactory.cs
using rin.Framework.Core;
using TerraFX.Interop.Vulkan;
using static TerraFX.Interop.Vulkan.Vulkan;

namespace rin.Framework.Graphics.Descriptors;

public class DescriptorLayoutFactory : Factory<VkDescriptorSetLayout, VkDescriptorSetLayoutCreateInfo, string>
{
    protected override void OnDispose(bool isManual)
    {
        var device = SGraphicsModule.Get().GetDevice();
        unsafe
        {
            foreach (var (_, layout) in GetData()) vkDestroyDescriptorSetLayout(device, layout, null);
        }
    }

    protected override string ToInternalKey(VkDescriptorSetLayoutCreateInfo key)
    {
        var result = ((int)key.flags).ToString();

        unsafe
        {
            for (var i = 0; i < key.bindingCount; i++)
            {
                var binding = key.pBindings[i];
                result +=
                    $"{binding.binding}${binding.descriptorCount}${(int)binding.descriptorType}{(int)binding.stageFlags}";
            }
        }

        return result;
    }

    protected override VkDescriptorSetLayout CreateNew(VkDescriptorSetLayoutCreateInfo key, string internalKey)
    {
        VkDescriptorSetLayout layout;
        unsafe
        {
            var device = SGraphicsModule.Get().GetDevice();
            vkCreateDescriptorSetLayout(device, &key, null, &layout);
        }

        return layout;
    }
}

File: lib/rin/rhi/DescriptorLayoutFactory.cpp
#include "rin/rhi/DescriptorLayoutFactory.h"

#include "rin/core/utils.h"
#include "rin/rhi/GraphicsModule.h"

namespace rin::rhi
{
    vk::DescriptorSetLayout DescriptorLayoutFactory::Create(const vk::DescriptorSetLayoutCreateInfo& key)
    {
        return GraphicsModule::Get()->GetDevice().createDescriptorSetLayout(key);
    }
    uint64_t DescriptorLayoutFactory::ToInternalKey(const vk::DescriptorSetLayoutCreateInfo& key)
    {
        uint64_t id{0};
        auto next = static_cast<const vk::DescriptorSetLayoutBindingFlagsCreateInfo*>(key.pNext);
        for (auto i = 0; i < key.bindingCount; i++)
        {
            auto binding = *(key.pBindings + i);

            id = hashCombine(
                id,
                binding.binding,
                binding.descriptorCount,
                static_cast<int>(binding.descriptorType),
                static_cast<uint32_t>(binding.stageFlags),
                static_cast<uint32_t>(*(next->pBindingFlags + i))
            );
        }

        return id;
    }
}
